crates/uv-publish/src/lib.rs

```diff
@@ -79,6 +81,13 @@ pub enum PublishSendError {
     RedirectError(Url),
 }
 
+pub trait Reporter: Send + Sync + 'static {
+    fn on_progress(&self, name: &str, id: usize);
+    fn on_download_start(&self, name: &str, size: Option<u64>) -> usize;
+    fn on_download_progress(&self, id: usize, inc: u64);
+    fn on_download_complete(&self);
+}
```

Comment:
How come you went for a trait here instead of asking for a concrete type? An enum would probably do the job in order to handle the case of the "dummy" reporter.

Reply:
I think this matches our other reporter patterns?

Reply:
Ah. That would be a candidate for un-genericizing code in a hope to improve compile times. :)

(It may result in little gain if there's only one instantiation of the generics outside of tests though.)

Reply:
It would be nice to have a single reporter crate that is shared instead of copying this code for every new feature.

Reply:
Yeah. I think it can be completely monomorphic too with no parametric polymorphism.